flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/committer/SinkCommitter.java

```diff
@@ -0,0 +1,463 @@
+  public static final String FLINK_JOB_ID = "flink.job-id";
+  public static final String OPERATOR_ID = "flink.operator-id";
+  private final String branch;
+  private final Map<String, String> snapshotProperties;
+  private final boolean replacePartitions;
+  private final int workerPoolSize;
+  private final String prefix;
+  private transient IcebergFilesCommitterMetrics committerMetrics;
+  private transient Table table;
+  private transient int maxContinuousEmptyCommits;
+  private transient ExecutorService workerPool;
+  private transient int continuousEmptyCheckpoints = 0;
+
+  public SinkCommitter(
+      TableLoader tableLoader,
+      String branch,
+      Map<String, String> snapshotProperties,
+      boolean replacePartitions,
+      int workerPoolSize,
+      String prefix,
+      IcebergFilesCommitterMetrics committerMetrics) {
+    this.branch = branch;
+    this.snapshotProperties = snapshotProperties;
+    this.replacePartitions = replacePartitions;
+    this.workerPoolSize = workerPoolSize;
+    this.prefix = prefix;
+    this.committerMetrics = committerMetrics;
+
+    if (!tableLoader.isOpen()) {
+      tableLoader.open();
+    }
+
+    this.table = tableLoader.loadTable();
+    this.maxContinuousEmptyCommits =
+        PropertyUtil.propertyAsInt(table.properties(), MAX_CONTINUOUS_EMPTY_COMMITS, 10);
+    Preconditions.checkArgument(
+        maxContinuousEmptyCommits > 0, MAX_CONTINUOUS_EMPTY_COMMITS + " must be positive");
+    this.workerPool =
+        ThreadPools.newWorkerPool("iceberg-worker-pool-" + table + "-" + prefix, workerPoolSize);
+    this.continuousEmptyCheckpoints = 0;
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<SinkCommittable>> commitRequests)
+      throws IOException, InterruptedException {
+    NavigableMap<Long, byte[]> manifestMap = Maps.newTreeMap();
+    NavigableMap<Long, CommitRequest<SinkCommittable>> commitRequestMap = Maps.newTreeMap();
+
+    if (commitRequests.isEmpty()) {
+      commitRequests.add(EMPTY_COMMIT_REQUEST);
+    }
+
+    for (CommitRequest<SinkCommittable> request : commitRequests) {
+      manifestMap.put(request.getCommittable().checkpointId(), request.getCommittable().manifest());
+      commitRequestMap.put(request.getCommittable().checkpointId(), request);
+    }
+
+    long maxCommittedCheckpointId = getMaxCommittedCheckpointId(commitRequestMap);
+
+    // Mark the already committed FilesCommittable(s) as finished
+    commitRequestMap
+        .headMap(maxCommittedCheckpointId, true)
+        .values()
+        .forEach(CommitRequest::signalAlreadyCommitted);
+
+    // Commit the remaining
+    SinkCommittable last = commitRequestMap.lastEntry().getValue().getCommittable();
+
+    NavigableMap<Long, byte[]> uncommitted =
+        Maps.newTreeMap(manifestMap).tailMap(maxCommittedCheckpointId, false);
+    if (!uncommitted.isEmpty()) {
+      commitUpToCheckpoint(uncommitted, last.jobId(), last.operatorId(), last.checkpointId());
```

Comment:
how do we notify commit success/failure to the `CommitRequest`?

```

        /**
         * The commit failed for known reason and should not be retried.
         *
         * <p>Currently calling this method only logs the error, discards the comittable and
         * continues. In the future the behaviour might be configurable.
         */
        void signalFailedWithKnownReason(Throwable t);

        /**
         * The commit failed for unknown reason and should not be retried.
         *
         * <p>Currently calling this method fails the job. In the future the behaviour might be
         * configurable.
         */
        void signalFailedWithUnknownReason(Throwable t);

        /**
         * The commit failed for a retriable reason. If the sink supports a retry maximum, this may
         * permanently fail after reaching that maximum. Else the committable will be retried as
         * long as this method is invoked after each attempt.
         */
        void retryLater();
```

Reply:
yeah. `CheckpointCommittableManagerImpl` marks those requests as committed if the commit doesn't throw an exception. but then why is `CommitRequest` even exposed to `Committer`. Let's say hypothetically IcebergCommitter wants to `signalFailedWithUnknownReason`. it will be ignored by the `CheckpointCommittableManagerImpl` which always call `CommitRequestImpl::setCommittedIfNoError` if no exception is thrown. A bit weird.

```
    @Override
    public Collection<CommittableWithLineage<CommT>> commit(
            boolean fullyReceived, Committer<CommT> committer)
            throws IOException, InterruptedException {
        Collection<CommitRequestImpl<CommT>> requests = getPendingRequests(fullyReceived);
        requests.forEach(CommitRequestImpl::setSelected);
        committer.commit(new ArrayList<>(requests));
        requests.forEach(CommitRequestImpl::setCommittedIfNoError);
        Collection<CommittableWithLineage<CommT>> committed = drainFinished();
        metricGroup.setCurrentPendingCommittablesGauge(() -> getPendingRequests(false).size());
        return committed;
    }
```